### PR TITLE
S3 Access

### PR DESCRIPTION
**Pull request by crstrong**
Moved the remote read/write from ImageData to RemoteConnection

### CODE REVIEW

File: src/TDBImage.cc [hunk 1]
```diff
@@ -79,6 +79,25 @@ TDBImage::TDBImage(const std::string &image_id) : TDBObject(image_id)
     _tile_order = false;
 }
 
+TDBImage::TDBImage(const std::string &image_id, RemoteConnection &connection) 
+    : TDBObject(image_id, connection)
```
[the remark this answers is not in the document]
Comment (luisremis) [reply src/TDBImage.cc]: I noticed the same happens on TDBObject.cc constructors.

File: src/TDBObject.cc [hunk 2]
```diff
@@ -58,13 +59,42 @@ TDBObject::TDBObject()
...
+        _ctx, "TileDB context initialization failed");
+
+    size_t pos = get_path_delimiter(image_id);
+
+    _group = get_group(image_id, pos);
+    _name = get_name(image_id, pos);
+
+    // set default values
+    _num_attributes = 1;
+    const char* attr = "value";
```
Comment (luisremis) [inline src/TDBObject.cc]: repeated code, same as the empty constructor. You can make a function here, or call the empty constructor.

File: SConstruct [hunk 3]
```diff
@@ -1,15 +1,25 @@
...
+                'src/TDBImage.cc', 
+                'src/RemoteConnection.cc',
+                'src/Exception.cc',
+                'src/utils.cc'
     ]
 
 env.SharedLibrary('libvcl.so', source_files,
-    LIBS = [ 'tiledb', 'opencv_core', 'opencv_imgproc', 'opencv_imgcodecs', 'gomp'],
+    LIBS = [ 'tiledb', 'opencv_core', 'opencv_imgproc', 'opencv_imgcodecs', 
+             'aws-cpp-sdk-core', 'aws-cpp-sdk-s3', 'gomp'],
```
Comment (luisremis) [inline SConstruct]: this libs should only be added if compiled with the enable=s3 flag.

File: src/ImageData.cc [hunk 4]
```diff
@@ -32,10 +32,10 @@
 #include <cstdio>
 #include <opencv2/imgproc.hpp>
 #include <iostream>
```
Comment (luisremis) [inline src/ImageData.cc]: iostream should not be here.

File: src/ImageData.cc [hunk 5]
```diff
@@ -231,6 +298,31 @@ ImageData::ImageData(const std::string &image_id)
     else
         _tdb = NULL;
 
+    _remote = NULL;
+}
+
+ImageData::ImageData(const std::string &image_id, RemoteConnection &connection)
```
Comment (luisremis) [inline src/ImageData.cc]: this is all repeated code. you should use the ImageData::ImageData(const std::string &image_id) constructor.

File: src/ImageData.cc [hunk 6]
```diff
@@ -82,19 +109,47 @@ ImageData::Write::Write(const std::string& filename, ImageFormat format,
       _metadata(metadata),
       _fullpath(filename)
 {
+    set_system();
+}
+
+void ImageData::Write::set_system()
```
Comment (luisremis) [inline src/ImageData.cc]: this function seems to be the same as Read::set_system. 
Can we move this function to the base class?

**Comment by luisremis** [comment]
shouldn't we consider adding the RemoteConnection class into VDMS (libcomm or some other place), as it is more general  not VCL-specific. @vishakha041 @philiplantz

**Comment by crstrong** [comment]
Closing this pull request, will open a new one

**Pull request closed without merging**